### PR TITLE
Fix final review comments from Leo

### PR DESCRIPTION
**Pull request by leo-barnes**
***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/av1-avif/pull/256.html" title="Last updated on Oct 16, 2024, 6:52 AM UTC (51a1b82)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/av1-avif/256/e0e048e...51a1b82.html" title="Last updated on Oct 16, 2024, 6:52 AM UTC (51a1b82)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -274,11 +276,11 @@ In general, it is recommended to use properties instead of [=Metadata OBUs=] in
...
+[[!AV1]] delegates the selection of which [=Operating Point=] to process to the application, by means of a function called <code>choose_operating_point()</code>. [=/AVIF=] defines the <code>[=OperatingPointSelectorProperty=]</code> to control this selection. In the absence of an <code>[=OperatingPointSelectorProperty=]</code> associated with an [=AV1 Image Item=], the [=/AVIF=] renderer is free to process any [=Operating Point=] present in the [=AV1 Image Item Data=]. In particular, <assert>when the [=AV1 Image Item Data=] is composed of a unique [=Operating Point=], the <code>[=OperatingPointSelectorProperty=]</code> should not be present</assert>. If an <code>[=OperatingPointSelectorProperty=]</code> is associated with an [=AV1 Image Item=], the <code>[=op_index=]</code> field indicates which [=Operating Point=] is expected to be processed for this item.
```
Comment (leo-barnes) [inline index.bs:279]: Regarding `AV1 Image Item` => `AV1 Image Item Data`, see issue #220.
In my mind, I see it like this:
1. The AV1 Image Item Data may contain one operating point or multiple operating points
2. The AV1 Image Item **always** consists of a unique operating point since the `a1op` defaults to selecting operating point 0 if not present.

So making statements about "when the AV1 Image Item is composed of a unique operating point" feels like something that is always true, but the text then says that if so the property should not be present.

Happy to revert if people disagree.

File: index.bs [hunk 2]
```diff
@@ -371,23 +373,27 @@ NOTE: The size of the last layer can be determined by subtracting the sum of the
...
 
 <p>An <dfn export>AV1 Depth Image Item</dfn> (respectively an <dfn export>AV1 Depth Image Sequence</dfn>) is an [=AV1 Auxiliary Image Item=] (respectively an [=AV1 Auxiliary Image Sequence=]), and as defined in [[!MIAF]], with the <code>[=aux_type=]</code> field of the <code>[=AuxiliaryTypeProperty=]</code> (respectively <code>[=AuxiliaryTypeInfoBox=]</code>) set to <code>urn:mpeg:mpegB:cicp:systems:auxiliary:depth</code>.</p>
 
 NOTE: [[!AV1]] supports encoding either 3-component images (whose semantics are given by the <code>[=matrix_coefficients=]</code> element), or 1-component images (monochrome). When an image requires a different number of components, multiple auxiliary images may be used, each providing additional component(s), according to the semantics of their <code>[=aux_type=]</code> field. In such case, the maximum number of components is restricted by number of possible items in a file, coded on 16 or 32 bits.
 
 <h3 id="derived-images">Derived Image Items</h3>
 
+<h4 id="grid-derivation">Grid Derived Image Item</h4>
+
+A <dfn noexport>grid derived image item</dfn> (<code>'[=grid=]'</code>) as defined in [[!HEIF]] may be used in an [=AVIF file=].
```
Comment (leo-barnes) [inline index.bs:386]: Added a small section on `grid` just so we can define it and refer to it in the rest of the text.

File: index.bs [hunk 3]
```diff
@@ -371,23 +373,27 @@ NOTE: The size of the last layer can be determined by subtracting the sum of the
...
+A tone map derived image item (<code>'[=tmap=]'</code>) as defined in [[!HEIF]] may be used in an [=AVIF file=]. <assert>When present, the base image item and the <code>'[=tmap=]'</code> image item should be grouped together by an '[=AVIF/altr=]' entity group as recommended in [[!HEIF]].</assert>
 
 <h4 id="sample-transform">Sample Transform Derived Image Item</h4>
 
-In these sections, a "sample" refers to the value of a pixel for a given channel.
+With a [=Sample Transform Derived Image Item=], pixels at the same position in multiple input image items can be combined into a single output pixel using basic mathematical operations. This can be used to work around codec limitations or for storing alterations to an image as non-destructive residuals. With a [=Sample Transform Derived Image Item=] it possible for [=/AVIF=] to support 16 or more bits of precision per sample, while still offering backward compatibility through a regular 8 to 12-bit [=AV1 Image Item=] containing the most significant bits of each sample.
```
Comment (leo-barnes) [inline index.bs:394]: Tried to make this description slightly more general so that it's clear that bit-depth extension is one use-case and others are possible.

File: index.bs [hunk 4]
```diff
@@ -986,191 +992,191 @@ The "Version(s)" column in the following table lists the version(s) of the boxes
...
         </tr>
         <tr>
             <td>&nbsp;</td>
             <td>dimg</td>
             <td>-</td>
-            <td>HEIF</td>
+            <td>[[!HEIF]]</td>
             <td>&nbsp;</td>
-            <td>Used when an image is derived from another image</td>
+            <td>Used when an image is [[#derived-images|derived from another image]]</td>
```
Comment (leo-barnes) [inline index.bs:1037]: FYI: I noticed in the Bikeshed spec that you can provide alternative text for links by using `[=xxx|This is my alt text=]`

File: index.bs [hunk 5]
```diff
@@ -1228,7 +1234,7 @@ Consider the following:
 
 This is equivalent to the following postfix notation (parentheses for clarity):
 
-<math><msub><mi>sample</mi><mi>output</mi></msub><mo>=</mo><mo>(</mo><mn>256</mn><mspace width="1ch"/><msub><mi>sample</mi><mn>1</mn></msub><mo>×</mo><mo>)</mo><msub><mi>sample</mi><mn>2</mn></msub><mo>+</mo></math>
+<math><msub><mi>sample</mi><mi>output</mi></msub><mo>=</mo><mo>(</mo><mn>256</mn><mspace width="0.5em"/><msub><mi>sample</mi><mn>1</mn></msub><mo>×</mo><mo>)</mo><mspace width="0.5em"/><msub><mi>sample</mi><mn>2</mn></msub><mo>+</mo></math>
```
Comment (leo-barnes) [inline index.bs:1237]: `"1ch"` does not seem to work in Safari. The space simply disappears. `"0.5em"` seems to work in both Chrome and Safari and gives roughly the same space.

File: index.bs [hunk 6]
```diff
@@ -469,11 +475,11 @@ aligned(8) class SampleTransform {
 
 The result of any computation underflowing or overflowing the intermediate bit depth is replaced by -2<sup><code>[=sato/num_bits=]</code>-1</sup> and 2<sup><code>[=sato/num_bits=]</code>-1</sup>-1, respectively. Encoder implementations should not create files leading to potential computation underflow or overflow. Decoder implementations shall check for computation underflow or overflow and clamp the results accordingly. Computations with [=sato/operands=] of negative values use the two’s-complement representation.
 
-<dfn noexport for="sato">token_count</dfn> is the expected number of [=sato/tokens=] to read.
+<dfn noexport for="sato">token_count</dfn> is the expected number of [=sato/tokens=] to read. The value of <code>[=sato/token_count=]</code> shall be greater than 0.
```
Comment (leo-barnes) [inline index.bs:478]: I need to wrap this `shall` in `<assert>` if we want to keep it.

File: index.bs (hunk 3 above)
Comment (leo-barnes) [reply index.bs:394]: Should change `This can be used` to `This can for example be used`

File: index.bs (hunk 2 above)
Comment (y-guyon) [reply index.bs:386]: Same as above: why not `[=grid derived image item=]` directly, with a matching HEIF dfn entry?

File: index.bs [hunk 7]
```diff
@@ -417,7 +423,7 @@ An expression is a series of [=sato/tokens=]. A [=sato/token=] is an [=sato/oper
 aligned(8) class SampleTransform {
     unsigned int(2) version = 0;
     unsigned int(4) reserved;
-    unsigned int(2) bit_depth; // Signed 8, 16, 32 or 64-bit.
+    unsigned int(2) bit_depth; // Enum signaling 8, 16, 32 or 64-bit.
```
Comment (y-guyon) [inline index.bs:426]: "Signed" was intended, as in not unsigned

File: index.bs [hunk 8]
```diff
@@ -469,11 +475,11 @@ aligned(8) class SampleTransform {
 
 The result of any computation underflowing or overflowing the intermediate bit depth is replaced by -2<sup><code>[=sato/num_bits=]</code>-1</sup> and 2<sup><code>[=sato/num_bits=]</code>-1</sup>-1, respectively. Encoder implementations should not create files leading to potential computation underflow or overflow. Decoder implementations shall check for computation underflow or overflow and clamp the results accordingly. Computations with [=sato/operands=] of negative values use the two’s-complement representation.
 
-<dfn noexport for="sato">token_count</dfn> is the expected number of [=sato/tokens=] to read.
+<dfn noexport for="sato">token_count</dfn> is the expected number of [=sato/tokens=] to read. <assert>The value of <code>[=sato/token_count=]</code> shall be greater than 0.</assert>
```
Comment (y-guyon) [inline index.bs:478]: This is indirectly covered by the single remaining stack element constraint. I am fine with making it explicit here.

Note: We considered naming it `token_count_minus1` but it seemed inconsistent with other similar boxes.

File: index.bs [hunk 9]
```diff
@@ -986,191 +992,191 @@ The "Version(s)" column in the following table lists the version(s) of the boxes
...
         </tr>
         <tr>
             <td>a1op</td>
             <td>&nbsp;</td>
             <td>-</td>
-            <td>AVIF</td>
+            <td>[=/AVIF=]</td>
             <td>ipco</td>
-            <td>Used to configure rendering of a multiple operating points image</td>
+            <td>Used to configure rendering of an image containing multiple operating points</td>
```
Comment (y-guyon) [inline index.bs:1165]: Based on your comment above, and assuming "image" here means "AV1 image item" and not "AV1 image item data", should it rather be:

> Used to configure the operating point to render an image with when there are multiple choices

**Review by y-guyon — APPROVED** [review]

File: index.bs (hunk 2 above)
Comment (leo-barnes) [reply index.bs:386]: See comment here https://github.com/AOMediaCodec/av1-avif/pull/256#discussion_r1793481437

File: index.bs [hunk 10]
```diff
@@ -251,15 +253,15 @@ The semantics of the clean aperture property (<code>'[=clap=]'</code>) as define
 
 <h4 id="other-item-property">Other Item Properties</h4>
 
-In addition to the Image Properties defined in this document, [=AV1 image items=] MAY also be associated with item properties defined in other specifications such as [[!HEIF]] and [[!MIAF]]. Examples of commonly used item properties are:
-    - <code>'[=colr=]'</code>
-    - <code>'[=pixi=]'</code>
+In addition to the Image Properties defined in this document, [=AV1 image items=] may also be associated with item properties defined in other specifications such as [[!HEIF]] and [[!MIAF]]. Examples of commonly used item properties are:
+    - <code>'[=colr=]'</code> (<dfn noexport for="AVIF">ColourInformationBox</dfn>)
```
[the remark this answers is not in the document]
Comment (y-guyon) [reply index.bs:257]: I feel it adds indirection. When one wants to look `ColourInformationBox` up, it will first lead to a place in AVIF, which itself redirects to HEIF. I get that this behavior is wanted to map `colr` to `ColourInformationBox`, but it also introduces confusion about which spec defines what.

> It's very hard to figure out how the various specs fit together.

I do not see an easy solution to this.

> It's unclear which things in HEIF and MIAF are actually supported/allowed in AVIF.

Solutions I can think of:
- Rely on the fact that everything that is not explicitly forbidden is allowed (~ current situation, could be written somewhere though)
- Succinctly list all boxes or features that are allowed (kind of an extended section 9. Box requirements) and declare all others as forbidden
- Create a new section for each supported box or feature, as was done for `tmap` and now `grid` etc. This feels like copy/pasting parts of ISOBMFF/HEIF/MIAF, but may be the clearest way for someone that just reads AV1-AVIF and not the inherited specs.

Example for `iden` which is allowed by AV1-AVIF but barely supported in implementations:
- Not explicitly forbidden hence allowed,
- Could be listed as `<code>[=iden=]</code>` and that's it,
- There could be a whole section about it.

Not sure which one is better than the others to be honest.

Anyway this seems beyond the topic of `AVIF/ColourInformationBox` and `AVIF/ColourInformationBox` barely answers that lack of clarity in my opinion.

I do not mind using both terms directly linked to 14496-12, especially if both are next to each other in the `dfn` list, but what about sticking to one of `colr` and `ColourInformationBox` if that is too much?

File: index.bs [hunk 11]
```diff
@@ -1201,9 +1207,9 @@ The "Version(s)" column in the following table lists the version(s) of the boxes
     - EDITORIAL: <a href="https://github.com/AOMediaCodec/av1-avif/pull/251">Clean up usage of dfn and linking</a>
     - <a href="https://github.com/AOMediaCodec/av1-avif/pull/250">Clarify required versions of non-essential item properties</a>
```
Comment (y-guyon) [inline index.bs:1208]: No `change-list` entry?

File: index.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (leo-barnes) [reply index.bs:257]: > Not sure which one is better than the others to be honest.

Yeah, that's what I was grappling with as well.

> I do not mind using both terms directly linked to 14496-12, especially if both are next to each other in the dfn list, but what about sticking to one of colr and ColourInformationBox if that is too much?

In order to improve consistency, how about this:
1. We switch all of the current usage of the full HEIF box names in AVIF to use the box types. That's mostly what we are using in any case, so it will only be a few changes. All of the box types that don't have explicit AVIF requirements or text links directly to the spec that defines it.
2. We add to the AVIF spec the boxes which are reasonably widely supported or something that we believe may end up being common in the foreseeable future. This means:
    1. For derived images we keep the sections for `grid` and `tmap` since they are widely supported. Since `iden` and `iovl` is only supported on Apple systems (I think?) it doesn't count as widely supported so we don't add sections for those. All use of `grid` and `tmap` inside of AVIF link to the AVIF sections which in turn links to HEIF.
    2. For item properties we add the properties that are widely supported or at least likely to be used to the list of boxes in `avif-required-boxes-additional`. 
    3. We either add common properties to the section "Other Item Properties" or completely remove that section and point it to `avif-required-boxes-additional`. I'm leaning towards the latter since it feels like we're duplicating stuff.

If we all agree on that, I guess the question comes down to which boxes to keeps in the list of boxes. 
- I added the `grpl`, `altr` and `ster` boxes since the first two are needed for `tmap` items and `ster` is pretty much the base requirement for stereo photos and we use them extensively for HEIC.
- I added all of the HDR item properties that have been added since they seem like they will likely be used now or in the future (and they are used in SlimHEIF).
- I also added the camera extrinsics and intrinsics since those are very important for stereo images and supported by libheif and the Apple platforms.

File: index.bs (hunk 5 above)
Comment (leo-barnes) [reply index.bs:1237]: Yeah, it's weird. Maybe it doesn't work correctly in <math> tags but works in other CSS stuff?

File: index.bs [hunk 12]
```diff
@@ -231,17 +233,17 @@ The syntax and semantics of the <dfn export>AV1ItemConfigurationProperty</dfn> a
...
 
     - In the absence of a <code>'[=lsel=]'</code> property associated with the item, or if it is present and its <code>[=layer_id=]</code> value is set to 0xFFFF:
-        - If no <code>[=OperatingPointSelectorProperty=]</code> is associated with the item, the <assert><code>'[=ispe=]'</code> shall document the dimensions of the last frame decoded when processing the operating point whose index is 0</assert>.
+        - If no <code>[=OperatingPointSelectorProperty=]</code> is associated with the item, the <assert><code>'[=ispe=]'</code> shall document the dimensions of the last frame decoded when processing the [=operating point=] whose index is 0</assert>.
```
[the remark this answers is not in the document]
Comment (leo-barnes) [reply index.bs:236]: Right, but I'm not sure if the term really needs to be capitalized. At least this way they all link to the same thing. Happy to capitalize all mentions if you prefer though.

File: index.bs [hunk 13]
```diff
@@ -221,8 +223,8 @@ The syntax and semantics of the <dfn export>AV1ItemConfigurationProperty</dfn> a
     - <assert>[=Sequence Header OBUs=] should not be present in the <code>[=AV1CodecConfigurationBox=]</code>.</assert>
     - <assert>If a [=Sequence Header OBU=] is present in the <code>[=AV1CodecConfigurationBox=]</code>, it shall match the [=Sequence Header OBU=] in the [=AV1 Image Item Data=].</assert>
     - <assert>The values of the fields in the <code>[=AV1CodecConfigurationBox=]</code> shall match those of the [=Sequence Header OBU=] in the [=AV1 Image Item Data=].</assert>
-    - <assert>The values of the bit depth and the number of channels derived from the <code>[=AV1CodecConfigurationBox=]</code> shall match the PixelInformationProperty if present.</assert>
-    - <assert>[=Metadata OBUs=], if present, shall match the values given in other item properties</assert>, such as the MasteringDisplayColourVolumeBox or ContentLightLevelBox.
+    - <assert>The values of the bit depth and the number of channels derived from the <code>[=AV1CodecConfigurationBox=]</code> shall match the <code>[=AVIF/PixelInformationProperty=]</code> if present.</assert>
```
[the remark this answers is not in the document]
Comment (leo-barnes) [reply index.bs:226]: Same as the other discussion on defining locally vs linking directly. Marking this as resolved so we keep the discussion to a single thread.

File: index.bs (hunk 1 above)
Comment (leo-barnes) [reply index.bs:279]: Hmmm... Good catch. Let me think while I write.

Section 2.2.2 says:
- No `lsel` or `lsel == 0xFFFF`
    - `a1op` missing => `ispe` maps to last frame in operating point 0
    - `a1op` present => `ispe` maps to last frame in operating point specified by `a1op`
- `lsel != 0xFFFF`
    - `a1op` missing => `ispe` maps to frame selected by `lsel` (operating point not actually specified)
    - `a1op` present => `ispe` maps to frame selected by `lsel` **in operating point specified by `a1op`**

For the case of `lsel == 0xFFFF`, the renderer is free to choose intermediate frames for progressive decoding, but it's fully defined what is considered the final output image. If `a1op` is present, only the layers in that operating point are used. **If `a10p` is not present, I think the renderer is technically allowed to choose any operating point as long as the final frame is the last frame in operating point 0.**

For the case of `lsel` being present and selecting an actual layer, I guess **the renderer is technically allowed to select any operating point that contains the layer if there is no `a1op` property.**

For the case of `lsel` not being present, the AV1 data is not supposed to be multilayered. **I think this means you can't really have multiple operating points.**

-----------

To summarize, I think I now understand why @cconcolato considered both the old and the new text equivalent. You **can** have an image item that has an unspecified operating point. The final output layer is fully defined, but that layer may belong to multiple operating points, and the writer does not technically need to specify which to use.

Given this I think I'll revert this change.

File: index.bs (hunk 2 above)
Comment (leo-barnes) [reply index.bs:386]: Marking as resolved so we keep the discussion to a single thread.

File: index.bs (hunk 7 above)
Comment (leo-barnes) [reply index.bs:426]: Good point. Will change to say it's an enum signaling xxxx

File: index.bs (hunk 9 above)
Comment (leo-barnes) [reply index.bs:1165]: Yeah, something like that would be better. Let me change it.

File: index.bs (hunk 11 above)
Comment (leo-barnes) [reply index.bs:1208]: Will add

File: index.bs (hunk 8 above)
Comment (leo-barnes) [reply index.bs:478]: Sounds good. Keeping it explicitly required here

File: index.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (y-guyon) [reply index.bs:257]: > 1. We switch all of the current usage of the full HEIF box names in AVIF to use the box types

So `colr` everywhere and no mention of `ColourInformationBox`. This is fine with me, although I would have picked the latter, personally.

> 2. We add to the AVIF spec the boxes which are reasonably widely supported

Sounds good to me, as long as it is explicit enough that they are not _defined_ by AVIF.

> 2.iii. completely remove "Other Item Properties" section

Agreed.

> which boxes to keep

I trust you on that. I am not that knowledgeable on what is used out there.

File: index.bs (hunk 5 above)
Comment (leo-barnes) [reply index.bs:1237]: Good question. That looks like it should be supported in Safari though. I pinged the WebKit folks to see if they know about it.

File: index.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (y-guyon) [reply index.bs:257]: Yes, I meant switching all, `colr` was just an example.

[the remark this answers is not in the document]
Comment (leo-barnes) [reply index.bs:257]: Actually, looking through the doc, we use the box names consistently in some cases (`FileTypeBox`). Let me handle this as follows instead:
For the cases where we _mostly_ refer to the box by the 4CC but at some point use the name, make it look like this:
```
<code>[=MasteringDisplayColourVolumeBox=]</code> (<code>'[=mdcv=]'</code>)
```

[the remark this answers is not in the document]
Comment (leo-barnes) [reply index.bs]: @y-guyon 
Let me know what you think. I've now made everything link directly to the base spec, _except in these places_:
1. `clap` in table of boxes links to `#clean-aperture-property` section
2. `grpl` in table of boxes links to `#groups` section
3. `altr` in table of boxes lins to `#altr-group` section
4. `ster` in table of boxes lins to `#ster-group` section

The places where we mention using `altr` groups now say something like this:
```
should be grouped together by an <code>'[=altr=]'</code> (see [[#altr-group]])
```

This way we link to the base spec _and_ add a link to the relevant section in AVIF.

File: index.bs (hunk 5 above)
Comment (leo-barnes) [reply index.bs:1286]: Turns out it's an issue in Firefox as well. WebKit folks opened an issue for it:
https://bugs.webkit.org/show_bug.cgi?id=281232

File: index.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (y-guyon) [reply index.bs]: Sorry, I missed your replies. This looks good, thank you for the changes.
